test/e2e/controller/etcd_test.go

```diff
@@ -0,0 +1,689 @@
+			peerTLSEnabledAfterUpdate:          true,
+			backupRestoreTLSEnabledAfterUpdate: true,
+		},
+		{
+			name:                      "enable-ptls-label-change",
+			peerTLSEnabledAfterUpdate: true,
+			labelsAfterUpdate: map[string]string{
+				"foo": "bar",
+			},
+		},
+		{
+			name:                               "enable-c-p-br-label-change",
+			clientTLSEnabledAfterUpdate:        true,
+			peerTLSEnabledAfterUpdate:          true,
+			backupRestoreTLSEnabledAfterUpdate: true,
+			labelsAfterUpdate: map[string]string{
+				"foo": "bar",
+			},
+		},
+		{
+			name:                                "disable-c-br-tls",
+			clientTLSEnabledBeforeUpdate:        true,
+			backupRestoreTLSEnabledBeforeUpdate: true,
+		},
+		// TODO: enable these once disabling peer TLS is supported
+		//{
+		//	name:                       "disable-p-tls",
+		//	peerTLSEnabledBeforeUpdate: true,
+		//},
+		//{
+		//	name:                                "disable-c-p-br-tls",
+		//	clientTLSEnabledBeforeUpdate:        true,
+		//	peerTLSEnabledBeforeUpdate:          true,
+		//	backupRestoreTLSEnabledBeforeUpdate: true,
+		//},
+		//{
+		//	name:                       "disable-ptls-label-change",
+		//	peerTLSEnabledBeforeUpdate: true,
+		//	labelsAfterUpdate: map[string]string{
+		//		"foo": "bar",
+		//	},
+		//},
+	}
+
+	g := NewWithT(t)
+
+	for _, provider := range providers {
+		for _, tc := range testCases {
+			tcName := fmt.Sprintf("update1-%s-%s", tc.name, provider)
+			t.Run(tcName, func(t *testing.T) {
+				t.Parallel()
+
+				testNamespace := testutils.GenerateTestNamespaceName(t, fmt.Sprintf("%s%s", testNamespacePrefix, tcName), 4)
+				logger := log.WithName(tcName).WithValues("etcdName", defaultEtcdName, "namespace", testNamespace)
+
+				initializeTestCase(g, testEnv, logger, testNamespace, defaultEtcdName)
+
+				logger.Info("running tests")
+				etcdBuilder := testutils.EtcdBuilderWithoutDefaults(defaultEtcdName, testNamespace).
+					WithReplicas(3).
+					WithDefaultBackup().
+					WithStorageProvider(provider, fmt.Sprintf("%s/%s", testNamespace, defaultEtcdName))
+
+				if tc.clientTLSEnabledBeforeUpdate {
+					etcdBuilder = etcdBuilder.WithClientTLS()
+				}
+				if tc.peerTLSEnabledBeforeUpdate {
+					etcdBuilder = etcdBuilder.WithPeerTLS()
+				}
+				if tc.backupRestoreTLSEnabledBeforeUpdate {
+					etcdBuilder = etcdBuilder.WithBackupRestoreTLS()
+				}
+				if tc.labelsBeforeUpdate != nil {
+					etcdBuilder = etcdBuilder.WithSpecLabels(tc.labelsBeforeUpdate)
+				}
+				etcd := etcdBuilder.Build()
+
+				logger.Info("creating Etcd")
+				testEnv.CreateAndCheckEtcd(g, etcd, timeoutEtcdReconciliation)
+				logger.Info("successfully created Etcd")
+
+				logger.Info("updating Etcd")
+				updateEtcdTLSAndLabels(etcd, tc.clientTLSEnabledAfterUpdate, tc.peerTLSEnabledAfterUpdate, tc.backupRestoreTLSEnabledAfterUpdate, tc.labelsAfterUpdate)
+				testEnv.UpdateAndCheckEtcd(g, etcd, timeoutEtcdReconciliation*2)
```

Comment:
This only checks for the matching number of replicas and the status/conditions, but doesn't check if the new labels are reflected in the statefulset or if TLS is actually enabled. For the labels, we can add an additional check to ensure the labels are what we expect and for TLS enablement, we can have a `peerTlsProber`/`backupRestoreTlsProber` of sorts to probe the etcd with the correct certificates and see if the request succeeds. We could also have a negative case where a non-TLS request succeeds before the change and fails afterward.